website/blog/2020-03-02-presto-lambda.md

```diff
@@ -0,0 +1,194 @@
+---
+title: Querying Nested Data with Lambda Functions
+author: Wenlei Xie
+authorURL: https://www.linkedin.com/in/wenleix/
+authorFBID: 681470066
+---
+
+Denormalized data with nested values (e.g. array/map) have become omnipresent in this Big Data era, as a lot of data naturally conforms to a nested representation [1, 2]. As a result it is important to provide an efficient and convenient way to query nested data. SQL traditionally does not include support for this.
+
+The pioneering work of Dremel proposed an extension to SQL based on recursive relational algebra to allow querying nested records [1], and is now available in BigQuery and the SQL:2016 standard. The following example shows how to transform array elements with this (adapted from [BigQuery Docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/arrays#creating-arrays-from-subqueries)):
+
+```sql
+SELECT elements,
+    ARRAY(SELECT v * 2
+          FROM UNNEST(elements) AS v) AS multiplied_elements
+FROM (
+    VALUES
+        (ARRAY[1, 2]),
+        (ARRAY[1, 3, 9]),
+        (ARRAY[1, 4, 16, 64])
+) AS t(elements)
+
+    elements    | multiplied_elements
+----------------+---------------------
+ [1, 2]         | [2, 4]
+ [1, 3, 9]      | [2, 6, 18]
+ [1, 4, 16, 64] | [2, 8, 32, 128]
+(3 rows)
+```
+
+While nested relational algebra provides an elegant and unified approach to query nested data, we found it could be challenging for users to track the “unnest stack” in mind when writing the query. In our experience, users are more comfortable to apply a given function (e.g lambda) to each element in the collection. This motivates us to introduce lambda expression into Presto to help query nested data, as illustrated below:
+
+```sql
+SELECT elements, 
+transform(elements, v -> v * 2) as multiplied_elements
+FROM (
+    VALUES
+        (ARRAY[1, 2]),
+        (ARRAY[1, 3, 9]),
+        (ARRAY[1, 4, 16, 64])
+) AS t(elements)
+```
+
+In Presto, lambda expression consists of argument list and lambda body, separated by `->`:
```

Comment:
lambda expression consists of argument list and lambda body -> a lambda expression consists of an argument list and lambda body